src/regexp_object.cpp

```diff
@@ -79,8 +127,150 @@ Value RegexpObject::match(Env *env, Value other, size_t start_index) {
     }
 }
 
+bool RegexpObject::has_match(Env *env, Value other, Value start) {
+    if (other->is_nil())
+        return false;
+    if (other->is_symbol())
+        other = other->as_symbol()->to_s(env);
+
+    other->assert_type(env, Object::Type::String, "String");
+    StringObject *str_obj = other->as_string();
+
+    nat_int_t start_index = 0;
+    if (start != nullptr) {
+        if (start.is_fast_integer()) {
+            start_index = start.get_fast_integer();
+        } else if (start->is_integer()) {
+            start_index = start->as_integer()->to_nat_int_t();
+        }
+    }
+    if (start_index < 0) {
+        start_index += str_obj->length();
+    }
+
+    OnigRegion *region = onig_region_new();
+    int result = search(str_obj->c_str(), start_index, region, ONIG_OPTION_NONE);
+
+    Env *caller_env = env->caller();
+
+    if (result >= 0) {
+        return true;
+    } else if (result == ONIG_MISMATCH) {
+        onig_region_free(region, true);
+        return false;
+    } else {
+        onig_region_free(region, true);
+        OnigUChar s[ONIG_MAX_ERROR_MESSAGE_LEN];
+        onig_error_code_to_str(s, result);
+        env->raise("RuntimeError", (char *)s);
+    }
+}
+
 Value RegexpObject::source(Env *env) {
     return new StringObject { pattern() };
 }
 
+Value RegexpObject::to_s(Env *env) {
+    StringObject *out = new StringObject { "(" };
+
+    auto is_m = options() & RegexOpts::MultiLine;
+    auto is_i = options() & RegexOpts::IgnoreCase;
+    auto is_x = options() & RegexOpts::Extended;
+
+    const char *str = pattern();
+    size_t len = strlen(str);
+    size_t start = 0;
+
+    if (str[start] == '(' && (start + 1) < len && str[start + 1] == '?' && str[len - 1]) {
+        /*
+        if there is only a single group fully-enclosing the regex, then
+        we won't need to wrap it all in another group to specify the options
+        */
+        bool active = true;
+        size_t i;
+        bool will_be_m = is_m;
+        bool will_be_i = is_i;
+        bool will_be_x = is_x;
+        for (i = start + 2; i < len && str[i] != ':'; i++) {
+            auto c = str[i];
+            switch (c) {
+            case 'm':
+                will_be_m = active;
+                break;
+            case 'i':
+                will_be_i = active;
+                break;
+            case 'x':
+                will_be_x = active;
+                break;
+            case '-':
+                if (!active) // this means we've already encountered a '-' which is illegal, so we just to append_options;
+                    goto append_options;
```

Comment:
:scream: I can't do the `goto`, sorry. Can we achieve the same result with a boolean `append_options = true` then falling through to subsequent code?

Reply:
Let me try cleaning this up, I might have an idea :)

Reply:
does this look any better now? 
Anyhow, the idea here was to stop parsing the options in the top-level group if the format was incorrect or the top level group wasn't the only top level group nor completely wrapping the regex :)